### PR TITLE
Code for initializing dynamo db in local

### PR DESCRIPTION
**Pull request by sarveshkaushal**
In an attempt to fix unit tests, I wrote initialization code for local dynamo instance at port 9000.
However i could not fix the unit test case(addReport_sanity), because it was trying to initialize actual object(ReportsDao). Should we use some kind of mocking framework? I am noob in Kotlin so detailed comments are welcome.

### CODE REVIEW

**Comment by sarveshkaushal** [comment]
Not sure why build is failing in the pipeline. It is running fine in my local.

**Comment by jmccance** [comment]
@sarveshkaushal Just pulled this locally and get the same errors as CI. Maybe something didn't get checked in/

```
./gradlew build

> Task :compileTestKotlin FAILED
e: /Users/joel/personal/coepi-backend-aws/src/test/kotlin/org/coepi/api/dao/ReportsDaoTest.kt: (6, 42): Unresolved reference: local
e: /Users/joel/personal/coepi-backend-aws/src/test/kotlin/org/coepi/api/dao/ReportsDaoTest.kt: (42, 26): Unresolved reference: ServerRunner

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':compileTestKotlin'.
> Compilation error. See log for more details

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 1s
5 actionable tasks: 1 executed, 4 up-to-date
```

image

**Comment by sarveshkaushal** [comment]
Thanks for input. I have added the dependencies in the build file. I guess i removed the build file from staged changes by mistake.

File: src/test/kotlin/org/coepi/api/dao/ReportsDaoTest.kt [hunk 1]
```diff
@@ -16,13 +22,49 @@ import java.util.*
...
+            System.setProperty("sqlite4java.library.path", "./build/libs")
+            val server = ServerRunner.createServerFromCommandLineArgs(arrayOf("-inMemory", "-port", "9000"))
+            server.start();
+
+            val dynamoClient = AmazonDynamoDBClientBuilder.standard()
+                    .withEndpointConfiguration(AwsClientBuilder.EndpointConfiguration(LOCAL_URL, REGION))
+                    .build();
+            val dynamoDb = DynamoDB(dynamoClient)
+
+            val table = dynamoDb.createTable(
```
Comment (bhushanRamnani) [inline src/test/kotlin/org/coepi/api/dao/ReportsDaoTest.kt:50]: You can use DDB Mapper and re-use the models to create table.

https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/services/dynamodbv2/datamodeling/DynamoDBMapper.html

File: src/test/kotlin/org/coepi/api/dao/ReportsDaoTest.kt [hunk 2]
```diff
@@ -16,13 +22,49 @@ import java.util.*
...
     val reportData = "foobar".toByteArray(Charset.defaultCharset())
 
+    companion object{
+
+        val LOCAL_URL= "http://localhost:9000";
+        val REGION = "us-west-2"
+
+        @BeforeAll
+        @JvmStatic
+        internal fun startDynamoDb(){
```
Comment (bhushanRamnani) [inline src/test/kotlin/org/coepi/api/dao/ReportsDaoTest.kt:40]: Can we make this re-usable for other test classes can use it? One way is to create an abstract DaoTestBase class which provides a method for local table creation.

Please add the support for TCNReportsDao as well.

File: build.gradle [hunk 3]
```diff
@@ -39,9 +46,22 @@ dependencies {
...
+    task copyNativeDeps(type: Copy) {
+        from configurations.dynamodb
+        into "$project.buildDir/libs/"
+    }
+
+    test.dependsOn copyNativeDeps
+    test.doFirst {
+        systemProperty "java.library.path", "./build/libs"
+    }
 }
```
Comment (bhushanRamnani) [inline build.gradle:66]: Please add the below line:

```
test {
    useJUnitPlatform()
}
```

Comment (sarveshkaushal) [reply build.gradle:69]: Done

File: src/test/kotlin/org/coepi/api/dao/ReportsDaoTest.kt (hunk 2 above)
Comment (sarveshkaushal) [reply src/test/kotlin/org/coepi/api/dao/ReportsDaoTest.kt]: Created a separate class to initialize dynamo db. Rather than an abstract class keeping it as a utility class.

File: src/test/kotlin/org/coepi/api/dao/ReportsDaoTest.kt (hunk 1 above)
Comment (sarveshkaushal) [reply src/test/kotlin/org/coepi/api/dao/ReportsDaoTest.kt]: Converted the code to create table using DDB Mapper.

File: src/test/kotlin/org/coepi/api/base/LocalDynamoDB.kt [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+                .build();
+        var ddbMapper = DynamoDBMapper(ddb);
+        createTable(ddb, ddbMapper)
+
+    }
+
+    /**
+     * TODO: Make it generic to create table based on class name passed to this method.
+     */
+    private fun createTable(ddb: AmazonDynamoDB, ddbMapper: DynamoDBMapper){
```
Comment (bhushanRamnani) [inline src/test/kotlin/org/coepi/api/base/LocalDynamoDB.kt:35]: Why do we need ReportsTableTest ? The idea is to unit test the table we use in the codebase.

File: src/test/kotlin/org/coepi/api/base/LocalDynamoDB.kt [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+        val request = ddbMapper.generateCreateTableRequest(ReportsTableTest::class.java)
+                .withProvisionedThroughput(ProvisionedThroughput(10, 10))
+        try{
+            ddb.createTable(request)
+            println("Created Reports table for test")
+            //fireTestQuery(ddb) // TODO: Invoke helper method to check the vales in the initialized table
+        } catch (ex: AmazonServiceException){
+            println(ex.errorMessage)
```
Comment (bhushanRamnani) [inline src/test/kotlin/org/coepi/api/base/LocalDynamoDB.kt:43]: this should be thrown

File: src/test/kotlin/org/coepi/api/base/ReportsTableTest.kt [hunk 6]
```diff
@@ -0,0 +1,13 @@
+package org.coepi.api.base
+
+import com.amazonaws.services.dynamodbv2.datamodeling.*
+
+@DynamoDBTable(tableName = "Report")
+data class ReportsTableTest(
```
Comment (bhushanRamnani) [inline src/test/kotlin/org/coepi/api/base/ReportsTableTest.kt:6]: Why is this needed ? We can simply use the table models that we already have.

File: src/test/kotlin/org/coepi/api/base/LocalDynamoDB.kt (hunk 5 above)
Comment (sarveshkaushal) [reply src/test/kotlin/org/coepi/api/base/LocalDynamoDB.kt:43]: Do you mean after catching the exception just throw it again??

Comment (bhushanRamnani) [reply src/test/kotlin/org/coepi/api/base/LocalDynamoDB.kt:43]: I mean why swallow the exception at all ? If the table creation fails, unit tests can't proceed.

Comment (sarveshkaushal) [reply src/test/kotlin/org/coepi/api/base/LocalDynamoDB.kt:43]: Ok. I can remove the try catch.

File: src/test/kotlin/org/coepi/api/base/LocalDynamoDB.kt (hunk 4 above)
Comment (sarveshkaushal) [reply src/test/kotlin/org/coepi/api/base/LocalDynamoDB.kt:35]: Report model will need some modifications for that. Some annotations like @DynamoDBTable(tableName="Report")
Also Hash key mapping for dynamo-mappe  to create it.

File: src/test/kotlin/org/coepi/api/base/LocalDynamoDB.kt (hunk 5 above)
Comment (sarveshkaushal) [reply src/test/kotlin/org/coepi/api/base/LocalDynamoDB.kt]: Keeping to TODO comments as is to address in future PRs.

File: src/test/kotlin/org/coepi/api/base/ReportsTableTest.kt (hunk 6 above)
Comment (sarveshkaushal) [reply src/test/kotlin/org/coepi/api/base/ReportsTableTest.kt]: It requires to modify the existing Report class.

**Comment by sarveshkaushal** [comment]
Need inputs on modifying Report class. Without adding few annotations, the table creation is failing in test.
